### DISCUSSION TITLE
Grafana integrating keycloak,  is it possible to skip login page completely?

### DISCUSSION THREAD

**Discussion by WalkerWalker** [keycloak/keycloak#19049] [Q&A]
I am asking almost the same question as described here https://stackoverflow.com/questions/48220504/login-to-keycloak-using-api

>I have 2 different applications: say **Application1** and **Application2**.
>I have integrated Application2 with keycloak and I am able to login to this application using Keycloak's login page.
>Now what I want is, if I login to my Application1 (without keycloak), I should be able to call some API of keycloak to login to application2 (without rendering keycloak's login page).
>It is feasible? If yes, how?
>Any help will be highly appreciated.
>Thanks

Only that application2 is grafana.   I want to login to application1 without keycloak and call some API of keycloak behind the scene so that when user goes to grafana, he/she doesn't need to go through any login page again.  Would that be possible?

PS.  Application1 is our legacy application and we decide to not develop anymore on the legacy backend.

**Comment by darius-m** [comment]
The first answer for the stackoverflow question appears to be pretty pertinent. If you trust Application1 enough, you could create a client inside keycloak that has the Direct Access Grant (aka resource owner password flow) enabled, which entrusts the application with the user's username and password. The application can authenticate the user normally and then simply send a POST request as described in the answer, that then creates the user's session in Keycloak.

However, a significant issue is that even if you can achieve this, it is likely impossible to not redirect the user through the Keycloak login page - or at least perform the authentication steps (POST request) inside the browser. This is because Keycloak must have *something* to authenticate the user when subsequent requests arrive. If you look at the [default browser flows](https://www.keycloak.org/docs/latest/server_admin/index.html#_authentication-flows), that something is either username and password (for interactive logins) or something like a cookie that must be saved in the browser. You could probably also use Kerberos, but that is likely out of scope.

The reason single sign-on platforms were created is because cookies (and thus user sessions) can only be accessed on the domain that they were created on. Since `domain2.com` cannot access the cookies set by `domain1.com`, `domain2.com` cannot verify that the user from `domain1.com` is authenticated. SSO platforms deal with this by separating user authentication and information brokerage. Whenever the user wants to authenticate on either `domain1.com` or `domain2.com`, they are redirected to `sso.com` that verifies their identity (using either the cookies it had set or the user credentials) and then returns a confirmation message (that may contain user information) or an error for the request. Since only `sso.com` deals with the user credentials, the user can authenticate only once, and the session cookies can be reused regardless of who initiated the authentication request, by redirecting the user to the SSO domain.

**Reply by WalkerWalker** [reply]
Thank you so much for the reply.  The answer even gives me hope, even though you are saying "not likely possible".  Let me provide a bit more information.  

The applicaiton1 and keycloak are at `domain1.com` (behind reverse-proxy).  The application2 (grafana) is at `domain2.com`.  Now when user attempts to visit grafana at `domain2.com`, he/she will be redirect-ed to keycloak login page at `domain1.com`, after login, be redirect-ed back to grafana at `domain2.com`.   If the user has previously already login-ed at `domain1.com` and received the cookies `KEYCLOAK_IDENTITY`, `KEYCLOAK-SESSION`, `AUTH_SESSION_ID` at `domain1.com`, then grafana at `domain2.com` will be logged in already.   You probably know all these already, how single sign on works.

After a little bit investigation,  I found out that as long as the cookie `KEYCLOAK_IDENTITY` is there at `domain1.com`,  grafana will be logged in. I also find that `KEYCLOAK_IDENTITY`  is essentially a jwt token. Therefore I am thinking/hoping about application1 getting this token through some keycloak endpoints and add it manually to the cookie. I have already achieved the answer described in the stack-overflow and successfully used it in another application (say application3).  But if I get the jwt token I received from calling that token endpoint, and manually set it as a cookie with key `KEYCLOAK_IDENTITY`,  grafana doesn't buy that.  So I must be missing something or my hope is too naive.  Can you provide some input?

**Reply by darius-m** [reply]
I'm not entirely sure that there is any way to set a sane `KEYCLOAK_IDENTITY` from an external client, since there is no reason to expect an external application (client) to ever access it. It is simply a token Keycloak can compare against the data stored in Infinispan to determine if the user's session is valid or not. It also sounds like it may become a hassle if Keycloak's behavior changes even slightly after an upgrade.

If you can modify your application's login process, I think you should rather add a POST request (redirect) through the user's browser that logs the user in Keycloak. That way you could rely on Keycloak to properly set its session cookies and perform whatever checks it needs. It would basically be the resource owner password flow, but you would not use the tokens returned by Keycloak after a valid sign-in request, so the checking the response is not required (basically avoiding most of OIDC's complexity).

**Reply by WalkerWalker** [reply]
Thank you for the reply.  I need to digest a bit of your reply.  

I cannot (don't want to) modify the legacy backend code of application1 but can modify the frontend code.  And we decide to use the existing frontend login page to lead, rather than the keycloak login page.  Therefore, after the "login" button is clicked,  in addition to login to the legacy backend, we also want to "simulate" a login to keycloak behind the scene with the username and password.  What I use now is calling a POST request to the `token` endpoint, `https://domain1.com/keycloak/realms/realm-id/protocol/openid-connect/token` and get a JWT token with it.   What you just described seems to be something I need exactly but have missed all along? 
> add a POST request (redirect) through the user's browser that logs the user in Keycloak.

**Reply by darius-m** [reply]
Does the javascript library that you use to send the POST request not also provide cookie access for the request? I would expect it to have the possibility to save cookies in a jar, that you can use later, and then you could simply save the cookies Keycloak had sent.

**Reply by WalkerWalker** [reply]
yes.  I have full access reading and writing the cookie.  To which endpoint should I send the POST request?   Right now I am sending to `https://domain1.com/keycloak/realms/realm-id/protocol/openid-connect/token`,  but should I use something else?

**Reply by darius-m** [reply]
Ah, your issue is that the token endpoint does not generate a cookie... That makes it quite a bit more difficult. You will probably have to perform a full authentication using the implicit flow (the more secure "code" flow may also work). This basically means that you would have to emulate the browser going to the regular login page, parse the `action` field of the login form, and then send a request that contains the username and password to that URL. The request going through the regular login steps should be able to provide you a working session cookie.

This is something you need to experiment with, since I have not done this (like this) before - I have done something similar for load testing, but a proper third party client was configured, and I simply had to fetch the responses from it and Keycloak and go follow the redirects. The only manual step was fetching the login form's `action` URL.

**Reply by WalkerWalker** [reply]
Thank you for the reply. Really appreciated it.  

I just manually parsed it with my eye.  Is it something like the following?
```
https://domain1.com/keycloak/realms/my_realm-id/login-actions/authenticate?session_code=JPt-3gXvgSiCAHBdRAlJ3ri8hUQ0SCnjHWzx7Ztgte4&amp;execution=8a356d99-2bfd-4e9a-84d1-68bef77cbc11&amp;client_id=my_client_id&amp;tab_id=-zws2cFQGho
```

Also I am just a bit surprised and confused now why for sso from keycloak to trust "I have logged in somewhere else", which I understand is a very strong statement, the user has to explicitly login via the keycloak frontend? Why login through api is not considered a reasonable way to login?

**Reply by darius-m** [reply]
That URL appears to have the correct format. I think that you could probably send an authentication request to it using a POST that contains the username and password, and the server should return an error or valid code / user token.

For the last part, the `token` endpoint that you send the request to for the direct access grant is indeed an API that allows you to login your users. However, since in that case the authentication request is normally sent by a backend without redirecting the browser to the SSO, I don't think Keycloak would return a valid cookie that could be saved in the user's browser. As such, there would be no way to create a persistent session in the SSO - the user **is** authenticated, but since there is no persistence, when the second application requests the authentication, the user will have to login again.
